bee-message/src/payload/data.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use bee_packable::{error::{PackPrefixError, UnpackPrefixError}, Packable, Packer, PackError, Unpacker, UnpackError, VecPrefix};
+
+use core::{fmt, convert::Infallible};
+
+#[derive(Debug)]
+pub enum DataPackError {
+    InvalidPrefixLength,
+}
+
+impl From<PackPrefixError<Infallible, u32>> for DataPackError {
+    fn from(_: PackPrefixError<Infallible, u32>) -> Self {
+        Self::InvalidPrefixLength
+    }
+}
+
+impl fmt::Display for DataPackError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::InvalidPrefixLength => write!(f, "Invalid prefix length for data"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum DataUnpackError {
+    InvalidPrefixLength,
+}
+
+impl From<UnpackPrefixError<Infallible, u32>> for DataUnpackError {
+    fn from(_: UnpackPrefixError<Infallible, u32>) -> Self {
+        Self::InvalidPrefixLength
+    }
+}
+
+impl fmt::Display for DataUnpackError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::InvalidPrefixLength => write!(f, "Invalid prefix length for data"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct DataPayload {
+    version: u8,
+    data: Vec<u8>,
+}
+
+impl DataPayload {
+    pub const KIND: u32 = 0;
+
+    pub fn new(version: u8, data: Vec<u8>) -> Self {
+        Self { version, data }
+    }
+
+    pub fn version(&self) -> u8 {
+        self.version
+    }
+
+    pub fn data(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+}
+
+impl Packable for DataPayload {
```

Comment:
Can't this one be derived?

I just realized that maybe I'm saying things you already now but haven't done because it is a draft :trollface:

Reply:
This can't be derived because of validations 🙁